[SR-2473] Support non-linked Homebrew packages pkgconfig

Bouke opened the pull request:
Some Homebrew packages cannot be linked, like OpenSSL. This patch adds support for packages that are not linked. It will query Homebrew about the installed version of that package, and add an additional pkg-config search path.

There was a discussion about this being a fallback. However Apple ships ``/usr/lib/pkgconfig/openssl.pc``, so the fallback wouldn't be triggered.

Review questions:
* `brew` is assumed to live on PATH -- is that ok? Otherwise, we might try the default location and use an environment variable (eg ``BREW_PREFIX``) as fallback?
* If the ``brew`` command could not be found (or produces an error), should that issue a warning on the console?
* How can we create tests for this? We shouldn't assume ``brew`` being available or certain packages to be installed.
* <s>If the JSON could not be parsed, should that issue a warning on the console?</s>

neonichu commented:
I don't think we should assume people have `brew` installed, so us not being able to find it should be silent.

Bouke commented:
In the Jira issue, @aciidb0mb3r proposes to use the ``brew info NAME --json=v1`` output. In my first version I did just that, however a user might have multiple versions of a library installed. This results in multiple versions being listed in the json output. However the json doesn't mention the version that the user has selected using ``brew switch NAME VERSION``.

In the commit I just added, I've removed the JSON parsing. Homebrew maintains symlinks under ``/usr/local/opt/(NAME)/lib/pkgconfig`` which will point to the correct version. This makes this implementation even more robust, as it doesn't have to parse the json.

There aren't any tests related to brew at the moment (besides json-serialization of Package). The code is quite trivial at the moment, but tests are always a good idea.

aciidb0mb3r commented on Sources/PackageLoading/Module+PkgConfig.swift:
```diff
@@ -78,7 +86,31 @@ private extension SystemPackageProvider {
+            } else {
+                return nil
+            }
+//            return AbsolutePath("\(brewPrefix)/opt/\(name)/lib/pkgconfig")
```
This is TODO?

Bouke replied:
It's an oversight from refactoring

aciidb0mb3r commented on Sources/PackageLoading/Module+PkgConfig.swift:
```diff
@@ -78,7 +86,31 @@ private extension SystemPackageProvider {
+            // to the latest version. Instead use the version as symlinked
+            // in /usr/local/opt/(NAME)/lib/pkgconfig.
+            let brewPrefix: String
+            if let prefix = getenv("BREW_PREFIX") {
```
Is this standard? I cant seem to find a reference

Bouke replied:
It's non-standard, to provide a way to use Homebrew from a non-default location when brew isn't on PATH.

aciidb0mb3r replied:
I don't think we should do this

Bouke replied:
I used this before: 7b82fae1ec170e4ac6a66d96a5f0968bcacf8f03, how about using that?

aciidb0mb3r commented on Sources/PackageLoading/Module+PkgConfig.swift:
```diff
@@ -78,7 +86,31 @@ private extension SystemPackageProvider {
+            let brewPrefix: String
+            if let prefix = getenv("BREW_PREFIX") {
+                brewPrefix = prefix
+            } else if let prefix = try? Utility.popen(["/usr/local/bin/brew", "--prefix"]).chomp() {
```
I think its ok to assume if brew is installed its in PATH

Bouke replied:
However; does brew need to be available from Xcode? Once the xcodeproj file is generated, I would expect that it doesn't need brew anymore. The libraries have already been resolved and the correct compiler flags already copied into the xcodeproj file. 

A valid reason for making sure brew is available from Xcode would be debugging `swift build` itself, but I'm not sure on how to achieve that.

aciidb0mb3r replied:
Yeah it is not needed after Xcode is generated but still I'd expect it to work unless I am missing something

aciidb0mb3r commented on Sources/Utility/PkgConfig.swift:
```diff
@@ -59,9 +59,9 @@ public struct PkgConfig {
     ///
     /// - throws: PkgConfigError
-    public init(name: String, fileSystem: FileSystem = localFileSystem) throws {
+    public init(name: String, additionalSearchPaths: [AbsolutePath] = [], fileSystem: FileSystem = localFileSystem) throws {
```
Can you documentation for this above the init method stating the priority in which this will be searched

aciidb0mb3r commented on Sources/PackageLoading/Module+PkgConfig.swift:
```diff
@@ -78,7 +86,23 @@ private extension SystemPackageProvider {
+            guard let brewPrefix = try? Utility.popen(["brew", "--prefix"]).chomp() else {
+                return nil
+            }
+            return AbsolutePath("\(brewPrefix)/opt/\(name)/lib/pkgconfig")
```
```swift
return AbsolutePath(brewPrefix).appending(components: "opt", name, "lib", "pkgconfig")
```

aciidb0mb3r commented on Sources/PackageLoading/Module+PkgConfig.swift:
```diff
@@ -78,7 +86,23 @@ private extension SystemPackageProvider {
+                return nil
+            }
+            return AbsolutePath("\(brewPrefix)/opt/\(name)/lib/pkgconfig")
+        default: return nil
```
Instead of default case, lets write all the cases so whenever a case is added we're forced to rethink about this method

aciidb0mb3r commented on Sources/Utility/PkgConfig.swift:
```diff
@@ -55,13 +55,19 @@ public struct PkgConfig {
+    /// It will search `fileSystem` for the pkg config file in the following order:
+    /// * Paths defined in `PKG_CONFIG_PATH` environment variable
+    /// * Paths defined in `additionalSearchPaths` argument
+    /// * Built-in search paths (see `PkgConfig.searchPaths`)
```
We should have a test proving this search order is respected

aciidb0mb3r commented:
LGTM

aciidb0mb3r commented:
@swift-ci please test

aciidb0mb3r commented:
@swift-ci please test

abertelrud commented:
Looks good to me.

Bouke commented:
Ok, great! I've removed the "WIP" tag from the title.
